src/python/Utils/Throttled.py

```diff
@@ -16,41 +36,130 @@
 
 class _ThrottleCounter(object):
     """
-    _ThrottleCounter class define throttle parameter and
+    _ThrottleCounter class defines throttle parameter and
     enter/exit methods to work with `with` context.
     """
 
     def __init__(self, throttle, user, debug=False):
         self.throttle = throttle
         self.user = user
-        self.debug = debug
+        if debug:
+            self.throttle.logger.setLevel(logging.DEBUG)
 
     def __enter__(self):
         "Define enter method for `with` context"
         ctr = self.throttle._incUser(self.user)
-        if self.debug:
-            msg = "Entering throttled function with counter %d for user %s" \
-                    % (ctr, self.user)
-            self.throttle.logger.debug(msg)
-        if ctr >= self.throttle.getLimit():
-            self.throttle._decUser(self.user)
+        msg = "Entering throttled function with counter %d for user %s" \
+                % (ctr, self.user)
+        self.throttle.logger.debug(msg)
+        if ctr > self.throttle.getLimit():
             msg = "The current number of active operations for this resource"
             msg += " exceeds the limit of %d for user %s" \
                    % (self.throttle.getLimit(), self.user)
-            raise cherrypy.HTTPError("500, Internal Server Error", msg)
+            raise cherrypy.HTTPError("429 Too Many Requests", msg)
 
     def __exit__(self, exc_type, exc_value, exc_traceback):
         "Define exit method for `with` context"
         ctr = self.throttle._decUser(self.user)
-        if self.debug:
-            msg = "Exiting throttled function with counter %d for user %s" \
-                  % (ctr, self.user)
-            self.throttle.logger.debug(msg)
+        msg = "Exiting throttled function with counter %d for user %s" \
+                % (ctr, self.user)
+        self.throttle.logger.debug(msg)
+
+class _ThrottleTimeCounter(object):
+    """
+    _ThrottleTimeCounter class defines time range throttled mechanism
+    """
+
+    def __init__(self, throttle, user, trange):
+        self.throttle = throttle
+        self.user = user
+        self.trange = trange
+
+    def __enter__(self):
+        "Define enter method for `with` context"
+        ctr = self.throttle._incUser(self.user, self.trange)
+        if ctr > self.throttle.getLimit():
+            self.throttle._decUser(self.user)
+            msg = "The current number of active operations for this resource"
+            msg += " exceeds the limit of %d for user %s in last %s sec" \
+                   % (self.throttle.getLimit(), self.user, self.trange)
+            raise cherrypy.HTTPError("429 Too Many Requests", msg)
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        "Define exit method for `with` context"
+        pass
+
+
+class UserThrottleTime(object):
+    """
+    UserThrottle class defines how to handle throttle time range based mechanism.
```

Comment:
It would be really useful to say what the "time range based mechanism" is.  I think you've captured it in the ticket, but it'll be a bit hard to find the ticket in 5 years when we need it next.

Personally, I remain a bit confused as to the benefit of the time-based mechanism compared to the original one, which intended (bugs notwithstanding) to limit the number of concurrent active queries.

What resources is it protecting?  What access patterns does it limit?  Looking at the code, it appears that with `trange = 2`, then a user can effectively launch as many heavyweight queries as they want in parallel if they match the heavy query with two short queries.  This is because each query start increases the counter by 1 and, if matched with two short queries at the beginning of the time range (so a total increment of 3), these lines:

```
step = int(trange - abs(time.time()-last_time))
self.users[user] -= step
```

will cause the counter to decrement by approximately 4, at which point the `last_time` variable is reset to `time.time()`.

Hence, it appears I can create an infinite stream of heavyweight blocking queries if I'm willing to do short queries in a separate thread to push down the counter by more than 1.

Why was this algorithm implemented as opposed to something more common such as an exponential moving average?

(Again, it's more useful to put the explanations in the class comments ... that's what started the long question in the first place!)

Reply:
I replied already below, but comment was not put in this thread.